### PR TITLE
LabelModel parameter clamping fix (fixes issue 1422)

### LINKED ISSUE #1422: LabelModel produces equal probability for labeled data
## Issue description

I am using snorkel for creating binary text classification training examples with 9 labeling functions. However, I find that for some data that are trained with the label model, they receive a probabilistic label of equal probabilities (i.e. `[0.5 0.5]`), despite that they only receive label of one class from the labeling functions (e.g. `[-1  0 -1 -1  0 -1 -1 -1 -1]`, so only class `0` or `ABSTAIN`), why is that?

Besides, I find that setting `verbose = True` when defining the `LabelModel` does not print  the logging information.

Lastly, if producing the label `[0.5 0.5]` is a normal behavior, then when filtering out unlabeled data points, they should be removed as well, because they do not contribute to the training (of a classifier), and if the classifier that does not support probablistic labels  (e.g. FastText), using `argmax` will always lead to class `0` (which is undesired).

## Code example/repro steps

Below I show my code of defining `LabelModel`:
```
print('==================================================')
print('Training label model...')
label_model = LabelModel(cardinality=2, verbose=True)
label_model.fit(L_train=L_train, n_epochs=10000, lr=0.001, log_freq=100, seed=1)
print('Done!')
```

Below I show some of the logs I print:
```
Check results for data in the training set:  a_text_in_the_training_set_but_I_removed_here
        * Output of L_train for this data point is: [-1  0 -1 -1  0 -1 -1 -1 -1]
        * Output of probs_train for this data point is: [0.5 0.5]
        * Output of probs_train_filtered for this data point is: [0.5 0.5]
        * Output of preds_train_filtered for this data point is: 0
```

## Expected behavior

I expect that if a datapoint receives label of only a single class, then it should not get equal probabilities of both classes when the label model is trained.

## System info

* How you installed Snorkel (conda, pip, source): pip
* Build command you used (if compiling from source):
* OS: Windows
* Python version: 3.7
* Snorkel version: 0.9
* Versions of any other relevant libraries:

### PR DESCRIPTION
**Pull request by ajratner**
## Description of proposed changes
_High level:_ This change makes the clamping threshold of `LabelModel` parameters a hyperparameter in `train_config`, and sets to a much smaller default value to avoid issues in sparse label setting.

_In more detail:_ The `LabelModel` learns the conditional probabilities of the labeling functions outputting certain labels given certain true (but unobserved or latent) true labels, e.g. `P(\lambda = x | Y = y)`. We also clamp these estimates to some min / max values just to guard against pathological errors... but as one might guess, when `P(\lambda = x | Y = y)` is really small- for example in the common sparse label setting where LFs mostly abstain- then clamping at too high of a minimum value messes everything up- we basically save clamped parameters that say these sparse LFs are equally likely to be wrong versus right! Luckily, this is fixed by changing the clamping parameter, which is now a kwarg in `LabelModel.fit`, and defaults to a much smaller value.

## Related issue(s)
- **Fixes #1422**
- Possibly addresses #1437 (TBD)

## Test plan
- Added a test (marked complex) that successfully (a) replicates the original error in #1422 and (b) solves it.

## Checklist
* [x] I have read the **CONTRIBUTING** document.
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.

### CODE REVIEW

File: snorkel/labeling/model/label_model.py [hunk 1]
```diff
@@ -46,6 +46,8 @@ class TrainConfig(Config):
         A random seed to initialize the random number generator with
     log_freq
         Report loss every this many epochs (steps)
+    mu_clamp_min
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: mu_clamp_eps

File: snorkel/labeling/model/label_model.py [hunk 2]
```diff
@@ -46,6 +46,8 @@ class TrainConfig(Config):
...
+        Restrict the learned conditional probabilities to [mu_clamp_min, 1-mu_clamp-min]
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: nit: typo - -> _

File: snorkel/labeling/model/label_model.py [hunk 3]
```diff
@@ -58,6 +60,7 @@ class TrainConfig(Config):
     prec_init: float = 0.7
     seed: int = np.random.randint(1e6)
     log_freq: int = 10
+    mu_clamp_min: float = 0.0001
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: do we expect anyone to touch this? if no, why 1e-4 and not e.g. 1e-8?

File: snorkel/labeling/model/label_model.py [hunk 4]
```diff
@@ -801,6 +803,14 @@ def fit(
             # Update learning rate
             self._update_lr_scheduler(epoch)
 
+        # Clamp learned parameters
+        # Note: If mu_clamp_min is set too high, e.g. in sparse settings where LFs
+        # mostly abstain, this will cause null results!
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: what are "null results"?

File: snorkel/labeling/model/label_model.py [hunk 5]
```diff
@@ -801,6 +803,14 @@ def fit(
...
+        self.mu.data = self.mu.clamp(  # type: ignore
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: Why do this here (and outside of the computation graph) instead of directly at inference time?

**Comment by henryre** [comment]
@ajratner also please change the PR name to something more informative since it will be the commit name

File: snorkel/labeling/model/label_model.py (hunk 3 above)
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: Orthogonal questions... pass for now :)

File: snorkel/labeling/model/label_model.py (hunk 5 above)
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: Regarding doing it here- I view this as part of our current procedure for learning / setting `mu`, therefore would go here.  Also safer if mu used in multiple sub-functions.

As for outside the computation graph- `torch.clamp` returns a tensor not a parameter, that's why I did this

File: snorkel/labeling/model/label_model.py (hunk 3 above)
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: Actually... @henryre thoughts on setting it to `None`, and then defaulting to picking something prop. to `1/n`?  Basically that's setting it to allow params to still be sensitive to LFs that only label a few of the `n` data points.

Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: See what I just committed

**Comment by ajratner** [comment]
@henryre ready for re-review

File: snorkel/labeling/model/label_model.py [hunk 6]
```diff
@@ -801,6 +803,20 @@ def fit(
...
+        # Note: If mu_eps is set too high, e.g. in sparse settings where LFs
+        # mostly abstain, this will result in learning conditional probabilities all
+        # equal to mu_eps (and/or 1 - mu_eps)!
+        # Note: Use user-provided value, else default to 1 / n', where n' is n rounded
+        # to the closest power of ten; this rounding is done to make it more obvious
+        # when the parameters have been clamped.
+        if self.train_config.mu_eps >= 0:
+            mu_eps = self.train_config.mu_eps
+        else:
+            mu_eps = min(0.01, 1 / 10 ** np.round(np.log10(self.n)))
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: Does `ceil` make more sense here? as is now: n = 300 -> mu_eps = 0.01

File: test/labeling/model/test_label_model.py [hunk 7]
```diff
@@ -355,14 +365,14 @@ def setUp(self) -> None:
...
-        """Test the LabelModel's estimate of P and Y."""
+    def test_label_model_basic(self) -> None:
+        """Test the LabelModel's estimate of P and Y on a simple synthetic dataset."""
         np.random.seed(123)
         P, Y, L = generate_simple_label_matrix(self.n, self.m, self.cardinality)
 
         # Train LabelModel
         label_model = LabelModel(cardinality=self.cardinality, verbose=False)
-        label_model.fit(L, n_epochs=200, lr=0.01, seed=123)
+        label_model.fit(L, n_epochs=1000, lr=0.01, seed=123)
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: why do we need more epochs now?

File: test/labeling/model/test_label_model.py [hunk 8]
```diff
@@ -371,9 +381,39 @@ def test_label_model(self) -> None:
         np.testing.assert_array_almost_equal(P, P_lm, decimal=2)
 
         # Test predicted labels
-        Y_lm = label_model.predict_proba(L).argmax(axis=1)
-        err = np.where(Y != Y_lm, 1, 0).sum() / self.n
-        self.assertLess(err, 0.1)
+        score = label_model.score(L, Y)
+        self.assertGreaterEqual(score["accuracy"], 0.9)
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: nice!

File: test/labeling/model/test_label_model.py [hunk 9]
```diff
@@ -371,9 +381,39 @@ def test_label_model(self) -> None:
...
+
+        # Test estimated LF conditional probabilities
+        P_lm = label_model._get_conditional_probs().reshape(
+            (self.m, self.cardinality + 1, -1)
+        )
+        np.testing.assert_array_almost_equal(P, P_lm, decimal=2)
+
+        # Test predicted labels *only on non-abstained data points*
+        Y_pred = label_model.predict(L, tie_break_policy="abstain")
+        idx, = np.where(Y_pred != -1)
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: nit: `idx, _` for readability

File: test/labeling/model/test_label_model.py [hunk 10]
```diff
@@ -371,9 +381,39 @@ def test_label_model(self) -> None:
...
+        acc = np.where(Y_pred[idx] == Y[idx], 1, 0).sum() / len(idx)
+        self.assertGreaterEqual(acc, 0.65)
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: this feels like a pretty low bar. what is the accuracy without the clamping patch?

File: test/labeling/model/test_label_model.py [hunk 11]
```diff
@@ -344,6 +344,16 @@ def test_warmup(self):
             lr_scheduler_config = {"warmup_steps": 1, "warmup_unit": "batches"}
             label_model.fit(L, lr_scheduler_config=lr_scheduler_config)
 
+    def test_set_mu_eps(self):
+        MU_EPS = 0.0123
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: nit: lowercase if not a module-level constant

File: snorkel/labeling/model/label_model.py [hunk 12]
```diff
@@ -58,6 +60,7 @@ class TrainConfig(Config):
     prec_init: float = 0.7
     seed: int = np.random.randint(1e6)
     log_freq: int = 10
+    mu_eps: float = -1.0
```
[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: I get a `mypy` error in my VSCode linter when I have it as `None`- is this something we have fixed in the config?

File: test/labeling/model/test_label_model.py (hunk 10 above)
Comment (ajratner) [reply test/labeling/model/test_label_model.py]: It's a hard problem when the LFs are this sparse; no way around that.  Without the clamping patch, the accuracy as measured here- not counting abstains- is roughly the same but coverage is far lower because of the error where LFs are counted as having null signal

Comment (henryre) [reply test/labeling/model/test_label_model.py]: Ok cool so `test_label_model_sparse` fails without this patch?

File: test/labeling/model/test_label_model.py (hunk 9 above)
Comment (ajratner) [reply test/labeling/model/test_label_model.py]: This doesn't work, there's only one element

Comment (henryre) [reply test/labeling/model/test_label_model.py]: ah ok, sg

File: test/labeling/model/test_label_model.py (hunk 10 above)
Comment (ajratner) [reply test/labeling/model/test_label_model.py]: Yes
